src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNNQuantizationStateWriter.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.KNN990Codec;
+
+import lombok.AllArgsConstructor;
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.store.IndexOutput;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Writes quantization states to off heap memory
+ */
+public class KNNQuantizationStateWriter {
+
+    private final IndexOutput output;
+    private List<FieldQuantizationState> fieldQuantizationStates = new ArrayList<>();
+
+    /**
+     * Constructor
+     * @param segmentWriteState segment write state containing segment information
+     * @throws IOException exception could be thrown while creating the output
+     */
+    public KNNQuantizationStateWriter(SegmentWriteState segmentWriteState) throws IOException {
+        String quantizationStateFileName = IndexFileNames.segmentFileName(
+            segmentWriteState.segmentInfo.name,
+            segmentWriteState.segmentSuffix,
+            KNNConstants.QUANTIZATION_STATE_FILE_SUFFIX
+        );
+
+        output = segmentWriteState.directory.createOutput(quantizationStateFileName, segmentWriteState.context);
+    }
+
+    /**
+     * Writes an index header
+     * @param segmentWriteState state containing segment information
+     * @throws IOException exception could be thrown while writing header
+     */
+    public void writeHeader(SegmentWriteState segmentWriteState) throws IOException {
+        CodecUtil.writeIndexHeader(output, "QuantizationCodec", 0, segmentWriteState.segmentInfo.getId(), segmentWriteState.segmentSuffix);
+    }
+
+    /**
+     * Writes a quantization state as bytes
+     * @param fieldName field name
+     * @param quantizationState quantization state
+     * @throws IOException could be thrown while writing
+     */
+    public void writeState(String fieldName, QuantizationState quantizationState) throws IOException {
+        byte[] stateBytes = quantizationState.toByteArray();
+        long position = output.getFilePointer();
+        output.writeBytes(stateBytes, stateBytes.length);
+        fieldQuantizationStates.add(new FieldQuantizationState(fieldName, stateBytes, position));
+    }
+
+    /**
+     * Writes index footer and other index information for parsing later
+     * @throws IOException could be thrown while writing
+     */
+    public void writeFooter() throws IOException {
+        long indexStartPosition = output.getFilePointer();
+        output.writeInt(fieldQuantizationStates.size());
+        for (FieldQuantizationState fieldQuantizationState : fieldQuantizationStates) {
+            output.writeString(fieldQuantizationState.fieldName);
```

Comment:
why we are writing the field name? we can write the ordinal of the field which will be just int and will save space for us.

Reply:
Does that still work for the use case where we want to read a specific field from a given segment?

Reply:
Yes it works. What you need to do is find the ordinal of the field and then use that ordinal to know where the field is present. I would recommend looking at some of the implementations of Lucene for flat vector writers. If we are not able to figure that out in that case using field_name is still fine.

Reply:
Looks like FieldInfo contains an attribute `number` that is the internal field number.  Would using that be better?